### ISSUE TITLE
List view throwing TypeError

### ISSUE THREAD

**Issue by curtisolson** [marmelab/react-admin#2055]
**What you were expecting:**
Just went through the tutorial (successfully) and tried to modify code to use my API. List view is not showing data and console has error. 

I was able to add the API call for login and on the list view I can see my API is successfully called. 

**What happened instead:**
Page shows title and "No results found"

Console error
```
index.js:2178 uncaught at handleFetch TypeError: newRecords.map is not a function
    at http://localhost:3001/static/js/bundle.js:69772:52
    at http://localhost:3001/static/js/bundle.js:69828:24
    at http://localhost:3001/static/js/bundle.js:69917:40
    at Array.reduce (<anonymous>)
    at ./node_modules/ra-core/lib/reducer/admin/resource/index.js.exports.default (http://localhost:3001/static/js/bundle.js:69914:30)
    at combination (http://localhost:3001/static/js/bundle.js:134671:29)
    at combination (http://localhost:3001/static/js/bundle.js:134671:29)
    at resettableAppReducer (http://localhost:3001/static/js/bundle.js:63329:16)
    at dispatch (http://localhost:3001/static/js/bundle.js:134907:22)
    at http://localhost:3001/static/js/bundle.js:119507:18
    at http://localhost:3001/static/js/bundle.js:132848:22
    at dispatch (http://localhost:3001/static/js/bundle.js:134462:18)
    at http://localhost:3001/static/js/bundle.js:134352:12
    at http://localhost:3001/static/js/bundle.js:133375:52
    at exec (http://localhost:3001/static/js/bundle.js:134015:5)
    at flush (http://localhost:3001/static/js/bundle.js:134056:5)
    at asap (http://localhost:3001/static/js/bundle.js:134029:5)
    at runPutEffect (http://localhost:3001/static/js/bundle.js:133372:69)
    at runEffect (http://localhost:3001/static/js/bundle.js:133321:307)
    at next (http://localhost:3001/static/js/bundle.js:133201:9)
    at currCb (http://localhost:3001/static/js/bundle.js:133274:7)
```

API Response Data from DevTools (Couldn't figure out how to format the data so it's a screen cap.)
<img width="1438" alt="screen shot 2018-07-20 at 9 44 20 am" src="https://user-images.githubusercontent.com/2302973/43014612-993cc470-8c01-11e8-99ed-c13c1e57414b.png">

API Response from Postman
```json
{
    "success": true,
    "data": [
        {
            "id": 1,
            "status": "DRAFT"
        },
        {
            "id": 2,
            "status": "DRAFT"
        },
        {
            "id": 3,
            "status": "DRAFT"
        },
        {
            "id": 4,
            "status": "READY"
        },
        {
            "id": 5,
            "status": "DRAFT"
        }
    ],
    "total": 5
}
```

Response Headers

```
Access-Control-Allow-Origin: * 
Access-Control-Expose-Headers: Content-Range 
Connection: keep-alive 
content-range: matches 0-5/5 
Date: Fri, 20 Jul 2018 16:33:12 GMT 
ETag: W/"a5-KZYtsf9f1aSutDA6i7vGzddXNZk" 
X-Powered-By: Express 
```

Matches.js code

```js
import React from 'react';
import { List, Edit, Create, Datagrid, ReferenceField, TextField, EditButton, DisabledInput, LongTextInput, ReferenceInput, SelectInput, SimpleForm, TextInput } from 'react-admin';

export const MatchList = (props) => (
    <List {...props}>
        <Datagrid>
            <TextField source="id" />
            <TextField source="status" />
        </Datagrid>
    </List>
);
```

The tutorial was using the dummy api and I switched to ra-data-simple-rest. I had to make a few changes to my API output to match the format expected, so I'm guessing this might be the source of the problem, but the error is not helpful. 

I originally had many columns and some were null, so I just left ID and Status. I do have 1 additional field returned in my results ("success": true). I removed that but got the same results.  

**Other information:**

**Environment**

* React-admin version: 2.1.2
* Last version that did not exhibit the issue (if applicable): N/A
* React version: 16.4.1
* Browser: Chrome/OSX
* Stack trace (in case of a JS error): above

**Comment by fzaninotto** [comment]
Hi, and thanks for your question. As explained in the [react-admin contributing guide](https://github.com/marmelab/react-admin/blob/master/.github/CONTRIBUTING.md), the right place to ask a "How To" question, get usage advice, or troubleshoot your own code, is [StackOverFlow](http://stackoverflow.com/questions/tagged/react-admin).

This makes your question easy to find by the core team, and the developer community. Unlike Github, StackOverFlow has great SEO, gamification, voting, and reputation. That's why we chose it, and decided to keep GitHub issues only for bugs and feature requests.

So I'm closing this issue, and inviting you to ask your question at:

http://stackoverflow.com/questions/tagged/react-admin

And once you get a response, please continue to hang out on the react-admin channel in StackOverflow. That way, you can help newcomers and share your expertise!

**Comment by curtisolson** [comment]
Happy to put it on stackoverflow. I figured since it was throwing an obscure error it qualified as a bug and not a question. The condition of the bug might be my API not being formatted properly, but the JS error seems like an unhandled exception. 

Excited to use your project and I'm focused on building an API and this library will help jump start the UI. Cheers.

**Comment by fzaninotto** [comment]
My 2c is that the `data` field of your response to a `GET_LIST` or `GET_MANY` request isn't an array - it should.

**Comment by fzaninotto** [comment]
Repoening as we can make the error more explicit.

**Comment by curtisolson** [comment]
The data is in an array. It's hard to see from the data in dev tools, but the Postman results show the array brackets.

`"data": 
    [
     { "id": 1, "status": "DRAFT" }, 
     { "id": 2, "status": "DRAFT" }, 
     { "id": 3, "status": "DRAFT" }, 
     { "id": 4, "status": "READY" }, 
     { "id": 5, "status": "DRAFT" } 
    ]`

**Comment by djhi** [comment]
Can you check your `dataProvider` doesn't return `{ data: { data: [...] } }` instead of `{ data: [...] }` ?

**Comment by curtisolson** [comment]
Here is my data from Postman. I removed whitespace for formatting.

{
    "success": true,
    "data": [
        {"id": 1, "status": "DRAFT"},
        {"id": 2, "status": "DRAFT"},
        {"id": 3, "status": "DRAFT"},
        {"id": 4, "status": "READY"},
        {"id": 5, "status": "DRAFT"}
    ],
    "total": 5
}

And here is the data from the docs. They look the same to me

https://marmelab.com/react-admin/DataProviders.html#response-format
// {
//     data: [
//         { id: 126, title: "allo?", author_id: 12 },
//         { id: 127, title: "bien le bonjour", author_id: 12 },
//         { id: 124, title: "good day sunshine", author_id: 12 },
//         { id: 123, title: "hello, world", author_id: 12 },
//         { id: 125, title: "howdy partner", author_id: 12 },
//     ],
//     total: 27
// }

**Comment by curtisolson** [comment]
Is there a public API that uses the ra-data-simple-rest? I could try that to see if I get the same error.

Looking for other ways I can troubleshoot. I've got several other parts of react-admin working. I've got auth working and can pass back the token to the API. I also have an add page working that will add records to the DB.

**Comment by fzaninotto** [comment]
[FakeRest](https://github.com/marmelab/FakeRest) uses the same REST format as ra-data-simple-rest. And [json-server](https://github.com/typicode/json-server) responses only differ by the name of the header for the content range (see ra-data-json-server for details).

Did you check the other verbs as I suggested in my previous message?

I'm going to close this issue anyway because we've improved the error messages in #2059. I invite you to switch to stackOverflow for troubleshooting.
